### ISSUE TITLE
`event.stopPropagation` is not working

### ISSUE THREAD

**Issue by chenzn1** [solidjs/solid#1786]
### Describe the bug

I used `event.stopPropagation` in the event in the application, but the event can still be executed outside the application.

### Your Example Website or App

https://codesandbox.io/p/sandbox/cool-thunder-mw586z?welcome=true

## Solidjs component
```
export defualt App() {
  return (
    <button
          style="margin-left: 20px"
          onclick={(event) => {
            console.log("solidjs click");
            event.stopPropagation();
          }}
        >
          click
        </button>
  )
}
```

## index.html
```
<body>
    <div id="root"></div>
    <script src="/src/index.jsx" type="module"></script>
  </body>
  <script>
    document.body.addEventListener('click', () => {
      console.log('click')
    })
  </script>
```

click button of solidjs app, the click event of `document.body` will execute

### Expected behavior

`event.stopPropagation` works fine.

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by wjq990112** [comment]
This behavior is different between native and delegate event:

https://playground.solidjs.com/anonymous/12059b7a-e53d-4e08-a13e-7be34da1b27d

**Comment by ryansolid** [comment]
Yeah, it is because all events are delegated from the document. The main reason for doing this is to support Portal delegation to be fair which could be placed anywhere in the document. The solution I suppose would be to keep track of all delegated events and re-apply them over portals. I think this is a good suggestion but will need some reasonable changes. Mostly making DOM Expressions Portal aware and changing how event delegation works. This will require at least a minor change if not major.

**Comment by nzdeep** [comment]
Perhaps I'm misunderstanding, but as I understand in this scenario then, shouldn't e.currentTarget be document? Currently it looks like currentTarget is faked to the button? (https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget)

**Comment by ryansolid** [comment]
The idea is that you can treat the delegated event as if it were not delegated within the system so we patch the currentTarget. This is a trick I got from Inferno and I imagine React does something similar as well.

**Comment by rnovacek** [comment]
Is there any workaround for this?

I have a link and I want to prevent the navigation and handle that in the component. But still want the link to be there so people can open the edit page in a new tab. I cannot figure out how to prevent the navigation.

```
<a
    href="/edit"
    onClick={(e) => {
        console.log('click', e)
        e.preventDefault();
        e.stopPropagation();
        setIsEditing((editing) => !editing);
        return false;
    }}
>
```

**Comment by rnovacek** [comment]
Ok, this seems to work:

```
on:click={(e: Event): boolean => {
    e.preventDefault();
    e.stopPropagation();
    setIsEditing((editing) => !editing);
    return false;
}}
```

But typescript doesn't like it:
```
Type '{ children: Element; href: string; class: string; "on:click": (e: Event) => boolean; }' is not assignable to type 'AnchorHTMLAttributes<HTMLAnchorElement>'.
  Property 'on:click' does not exist on type 'AnchorHTMLAttributes<HTMLAnchorElement>'. Did you mean 'onclick'?
```

**Comment by trusktr** [comment]
I too have had gripes with this. It would be nice if JSX were simply sugar for DOM as much as possible. Delegation should be an opt-in feature.

For example, if anything, `onclick={}` should be the pure native event handler directly on the element, and `on:click={}` with special syntax could be the opt-in delegation form.

I also don't see how this has anything to do with Portals: if we have the reference to the element, we add an event listener to it, and it would simply work, no matter where in the DOM it is located.

I think Solid is trying to do too much in this case. *Maybe* delegation solves some performance aspects that are in fact negligible, for sake of a benchmark result, but a delegation opt-in should easily solve that. In my long experience with web dev, individual event handlers have never been an issue, but I *would* delegate if needed (f.e. the familiar table with thousands of rows, but that's a special case, and defaulting to delegation everywhere just for that special case is overkill I think).

**Comment by yume-chan** [comment]
> I also don't see how this has anything to do with Portals:

Event delegation let events propagate through the JSX tree, not the DOM tree, for example:

https://playground.solidjs.com/anonymous/f787bc00-e21e-4ea8-8f57-7eca6d931e72

```jsx
function Counter() {
  return (
    <div onKeyDown={() => console.log("portal key down")} onKeyPress={() => console.log("portal key press")}>
      <Portal>
        <div onKeyDown={() => console.log("container key down")} onKeyPress={() => console.log("container key press")}>
          <input onKeyDown={() => console.log("input key down")} onKeyPress={() => console.log("input key press")} />
        </div>
      </Portal>
    </div>
  );
}
```

`input` is not a DOM descendant of the outermost `div`, without event delegation, events trigger from the `input` won't bubble to the outermost `div`. In this example, `keydown` is delegated, but `keypress` isn't. Notice that when you press a key on keyboard in the `input`, it prints "portal key down" but not "portal key press".

That's why Ryan was talking about

> The solution I suppose would be to keep track of all delegated events and re-apply them over portals.

We can add event handlers to every Portal, to dispatch events back to their JSX tree position, without relying on event delegation for basic event dispatching.

---

IMO, event delegation makes working with Portals easier, it definitely has its benefits. But, as Solid is a lower-level DOM framework, users might more frequently use it alongside with native DOM operations, including event handlers. Or maybe users use a third-party library which adds event handlers without knowing Solid's event delegation. Both situations can cause hard to debug bugs.

---

Anyway, it's possible to disable event delegation (except for Solid Start), although not very easily. First you need to disable it when compiling components:

```js
import solid from "vite-plugin-solid";
import { defineConfig } from "vite";

export default defineConfig({
  plugins: [
    solid({
      delegateEvents: false,
    }),
  ],
});
```

Then to disable `Dynamic` from using event delegation, add this to the entry point:

```js
import { DelegatedEvents } from "solid-js/web";

DelegatedEvents.clear();
```

Solid Start's Vite plugin doesn't follow the `delegateEvents: false` config for route pages:

* https://github.com/solidjs/solid-start/pull/1029

**Comment by this-spring** [comment]
I this the root node for delegated events should be on the root(id is app div) instead of document.

**Comment by KalyDanil** [comment]
Using ```onClickCapture = (e) => e.stopPropagation()``` instead of ```onClick = (e) => e.stopPropagation()``` help me

**Comment by ryansolid** [comment]
> Using `onClickCapture = (e) => e.stopPropagation()` instead of `onClick = (e) => e.stopPropagation()` help me

We don't have `onClickCapture`. If you want to do a capture event you can pass an object to the non-delegated event syntax.

```js
<div on:click={{
  handleEvent(e) {
    // your handler
    e.stopPropagation()
  },
  capture: true
}} />
```
‐---------

In general given the direction of things I don't see delegation by default going away. We can try to scope them down if we can do smarter stuff with Portals. But future hydration features makes delegation essential.

**Comment by ryansolid** [comment]
This is working as designed for delegated events. Solid delegates onClick at the document level, so stopPropagation() inside that handler cannot prevent earlier native bubbling listeners from having already run. Use on:click for native event listener semantics. We may revisit scoped delegation/portal event forwarding in a future major design, but there is no planned bug fix here. And with having reviewed this again with 2.0, this isn't going to be that future time.

The final repro is also expected: a native target listener can stop propagation before the event reaches Solid’s delegated document listener, so the delegated onClick will not run. Use on:click for a native element listener when ordering relative to other native listeners matters.
